[SPARK-25153][SQL] Improve error messages for columns with dots/periods

jeff303 opened the pull request:
### What changes were proposed in this pull request?

Check schema fields to see if they contain the exact column name, add to error message in DataSet#resolve

Add test for extra error message piece

Adds an additional check in `DataSet#resolve`, in the else clause (i.e. column not resolved), that appends a suffix to the error message for the `AnalysisException` if that column name is literally found in the schema fields, to suggest to the user that it might need to be quoted via backticks.

### Why are the changes needed?

Forgetting to quote such column names is a common occurrence for new Spark users.

### Does this PR introduce any user-facing change?

No (other than the extra suffix on the error message).

### How was this patch tested?

`test` was run for `core` in `sbt`, and passed.

jeff303 committed:
[SPARK-25153][SQL] Improve error messages for columns with dots/periods

Check schema fields to see if they contain the exact column name, add to error message in DataSet#resolve

Add test for extra error message piece

AmplabJenkins commented:
Can one of the admins verify this patch?

AmplabJenkins commented:
Can one of the admins verify this patch?

HyukjinKwon commented:
ok to test

HyukjinKwon commented:
Is it a duplicate of https://github.com/apache/spark/pull/22208? Please address the concerns listed there.

SparkQA commented:
**[Test build #110702 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/110702/testReport)** for PR 25807 at commit [`54b0b3b`](https://github.com/apache/spark/commit/54b0b3b466ebd0f158229a67d2988e02c430aabd).

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder-K8s/15865/
Test PASSed.

SparkQA commented:
**[Test build #110702 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/110702/testReport)** for PR 25807 at commit [`54b0b3b`](https://github.com/apache/spark/commit/54b0b3b466ebd0f158229a67d2988e02c430aabd).
 * This patch **fails Spark unit tests**.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test FAILed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/110702/
Test FAILed.

jeff303 commented:
I have updated this to use the `resolver` as per #22208, and modified the test to iterate both case sensitivity modes.  Of course, suggestions on the wording of this suffix are also welcomed.

SparkQA commented:
**[Test build #110795 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/110795/testReport)** for PR 25807 at commit [`ed4d2a0`](https://github.com/apache/spark/commit/ed4d2a0680d9caadbc2bf7050f59e2a3b3f1b4b1).

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder-K8s/15929/
Test PASSed.

AmplabJenkins commented:
Merged build finished. Test FAILed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/110795/
Test FAILed.

shaneknapp commented:
test this please

SparkQA commented:
**[Test build #110803 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/110803/testReport)** for PR 25807 at commit [`ed4d2a0`](https://github.com/apache/spark/commit/ed4d2a0680d9caadbc2bf7050f59e2a3b3f1b4b1).

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder-K8s/15936/
Test PASSed.

SparkQA commented:
**[Test build #110803 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/110803/testReport)** for PR 25807 at commit [`ed4d2a0`](https://github.com/apache/spark/commit/ed4d2a0680d9caadbc2bf7050f59e2a3b3f1b4b1).
 * This patch passes all tests.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/110803/
Test PASSed.

HyukjinKwon commented:
cc @dongjoon-hyun

dongjoon-hyun commented:
Oh, sorry for missing the ping here, @HyukjinKwon . I'll review today~

dongjoon-hyun commented:
Retest this please.

dongjoon-hyun commented on sql/core/src/main/scala/org/apache/spark/sql/Dataset.scala:
```diff
@@ -256,8 +256,13 @@ class Dataset[T] private[sql](
-          s"""Cannot resolve column name "$colName" among (${schema.fieldNames.mkString(", ")})""")
+        val fields = schema.fieldNames
+        val extraMsg = if (fields.exists(sparkSession.sessionState.analyzer.resolver(_, colName))) {
+          s"; did you mean to quote the `${colName}` column?"
```
nit. `${colName}` -> `$colName`

jeff303 replied:
Done

SparkQA commented:
**[Test build #111105 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/111105/testReport)** for PR 25807 at commit [`ed4d2a0`](https://github.com/apache/spark/commit/ed4d2a0680d9caadbc2bf7050f59e2a3b3f1b4b1).

dongjoon-hyun commented on sql/core/src/test/scala/org/apache/spark/sql/DatasetSuite.scala:
```diff
@@ -1841,6 +1842,24 @@ class DatasetSuite extends QueryTest with SharedSparkSession {
+        val errorMsg = intercept[AnalysisException] {
+          ds(colName)
+        }
+        assert(errorMsg.getMessage.contains(s"did you mean to quote the `${colName}` column?"))
```
nit. `${colName}` -> `$colName`

jeff303 replied:
Done

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder-K8s/16187/
Test PASSed.

dongjoon-hyun commented on sql/core/src/main/scala/org/apache/spark/sql/Dataset.scala:
```diff
@@ -256,8 +256,13 @@ class Dataset[T] private[sql](
-        throw new AnalysisException(
-          s"""Cannot resolve column name "$colName" among (${schema.fieldNames.mkString(", ")})""")
+        val fields = schema.fieldNames
+        val extraMsg = if (fields.exists(sparkSession.sessionState.analyzer.resolver(_, colName))) {
```
Please declare `val resolver = sparkSession.sessionState.analyzer.resolver` before line 257 because this is used twice.

jeff303 replied:
Done

dongjoon-hyun commented:
It would be good to ping @holdenk since she is the reporter of this issue.

dongjoon-hyun commented:
BTW, thank you for making a PR, @jeff303 .

SparkQA commented:
**[Test build #111105 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/111105/testReport)** for PR 25807 at commit [`ed4d2a0`](https://github.com/apache/spark/commit/ed4d2a0680d9caadbc2bf7050f59e2a3b3f1b4b1).
 * This patch **fails due to an unknown error code, -9**.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test FAILed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/111105/
Test FAILed.

SparkQA commented:
**[Test build #111225 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/111225/testReport)** for PR 25807 at commit [`7af8ad7`](https://github.com/apache/spark/commit/7af8ad7b8ba0da349c10a828cf203d64ced581ed).

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder-K8s/16300/
Test PASSed.

SparkQA commented:
**[Test build #111227 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/111227/testReport)** for PR 25807 at commit [`0416f04`](https://github.com/apache/spark/commit/0416f04bad9a64f967ed41f257021a9246b5f573).

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder-K8s/16302/
Test PASSed.

holdenk commented:
Thanks for making the PR @jeff303 & thanks for pinging me @dongjoon-hyun . I'll take a look at this tonight.

SparkQA commented:
**[Test build #111225 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/111225/testReport)** for PR 25807 at commit [`7af8ad7`](https://github.com/apache/spark/commit/7af8ad7b8ba0da349c10a828cf203d64ced581ed).
 * This patch passes all tests.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/111225/
Test PASSed.

SparkQA commented:
**[Test build #111227 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/111227/testReport)** for PR 25807 at commit [`0416f04`](https://github.com/apache/spark/commit/0416f04bad9a64f967ed41f257021a9246b5f573).
 * This patch passes all tests.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/111227/
Test PASSed.

holdenk reviewed (approved):
This looks good to me pending CI. I believe the JDK11 failure is unrelated. Let's see if Jenkins retest this please is still the magic words to use.

holdenk commented:
Also can someone clarify the GH actions builds - do we block on those now?

jeff303 commented:
Looks like the JDK 11 build failed with an issue resolving an artifact.

`[ERROR] Failed to execute goal on project spark-tools_2.12: Could not resolve dependencies for project org.apache.spark:spark-tools_2.12:jar:3.0.0-SNAPSHOT: Failed to collect dependencies at org.clapper:classutil_2.12:jar:1.1.2: Failed to read artifact descriptor for org.clapper:classutil_2.12:jar:1.1.2: Could not transfer artifact org.clapper:classutil_2.12:pom:1.1.2 from/to central (https://repo.maven.apache.org/maven2): Connection timed out (Read failed) -> [Help 1]`

holdenk commented:
LGTM merged to master
